amethyst_ui/src/prefab.rs

```diff
@@ -513,56 +637,70 @@ type UiPrefabData<A = AudioFormat, I = TextureFormat, F = FontFormat> = (
 /// - `A`: `Format` used for loading sounds
 /// - `I`: `Format` used for loading `Texture`s
 /// - `F`: `Format` used for loading `FontAsset`
-pub type UiPrefab<A = AudioFormat, I = TextureFormat, F = FontFormat> =
-    Prefab<UiPrefabData<A, I, F>>;
+pub type UiPrefab<
+    A = AudioFormat,
+    I = TextureFormat,
+    F = FontFormat,
+    CD = <NoCustomUi as ToNativeWidget>::PrefabData,
+> = Prefab<UiPrefabData<A, I, F, CD>>;
 
 /// Ui format.
 ///
 /// Load `UiPrefab` from `ron` file.
-#[derive(Clone, Debug, Serialize, Deserialize)]
-pub struct UiFormat;
+#[derive(Serialize, Deserialize, Derivative)]
+#[derivative(Clone(bound = ""), Debug(bound = ""), Default(bound = ""))]
+pub struct UiFormat<C>(PhantomData<C>);
 
-impl<A, I, F> SimpleFormat<UiPrefab<A, I, F>> for UiFormat
+impl<A, I, F, C> SimpleFormat<UiPrefab<A, I, F, C::PrefabData>> for UiFormat<C>
 where
     A: Format<Audio, Options = ()> + Sync + DeserializeOwned,
     I: Format<Texture, Options = TextureMetadata> + Sync + DeserializeOwned + Clone,
     F: Format<FontAsset, Options = ()> + Sync + DeserializeOwned + Clone,
+    C: ToNativeWidget<A, I, F> + for<'de> serde::Deserialize<'de>,
 {
     const NAME: &'static str = "Ui";
     type Options = ();
 
-    fn import(&self, bytes: Vec<u8>, _: ()) -> AssetResult<UiPrefab<A, I, F>> {
+    fn import(&self, bytes: Vec<u8>, _: ()) -> AssetResult<UiPrefab<A, I, F, C::PrefabData>> {
         use ron::de::Deserializer;
         use serde::Deserialize;
         let mut d =
             Deserializer::from_bytes(&bytes).chain_err(|| "Failed deserializing Ron file")?;
-        let root: UiWidget<A, I, F> =
+        let root: UiWidget<A, I, F, C> =
             UiWidget::deserialize(&mut d).chain_err(|| "Failed parsing Ron file")?;
         d.end().chain_err(|| "Failed parsing Ron file")?;
 
         let mut prefab = Prefab::new();
-        walk_ui_tree(root, 0, &mut prefab);
+        walk_ui_tree(root, 0, &mut prefab, Default::default());
```

Comment:
If you have to put default everywhere, doesn't that indicate that it should be taking Option?

Reply:
It could use `Option<ToNativeWidget::PrefabData>`, but that would mean that:
- in order to say that not custom data is used, custom UI implementation would have to use `ToNativeWidget::PrefabData = !` (never type is unstable, custom `enum Never {}` would need to implement `for<'a> PrefabData<'a>` and others which increases boilerplate) where in current form it may use `ToNativeWidget::PrefabData = ()`
- there is no way how to flag native UI defined in prefab (now you can use `default` to do it)

Implementation can use  `ToNativeWidget::PrefabData = Option<...>`.